File: blockchainPAdic/blockchainPAdic.py
import hashlib
import random
import json

def hash_data(data):
    """Erzeugt einen SHA-256 Hash aus Daten."""
    return hashlib.sha256(json.dumps(data, sort_keys=True).encode()).hexdigest()

def generate_node(level_name, level_count, parent_hash=""):
    """Erstellt einen Knoten mit 1-3 Kindern rekursiv."""
    node = {
        "name": f"{level_name}_{random.randint(1,1000)}",
        "hash": "",
        "children": []
    }
    # zufällige Anzahl von Kindern zwischen 1 und 3
    num_children = random.randint(1,3) if level_count > 0 else 0
    
    for _ in range(num_children):
        if level_count == 3:
            # Ebene Länder -> Bundesländer
            child = generate_node("Bundesland", level_count-1, parent_hash)
        elif level_count == 2:
            # Ebene Bundesländer -> Bezirke
            child = generate_node("Bezirk", level_count-1, parent_hash)
        elif level_count == 1:
            # Ebene Bezirke -> Dörfer
            child = generate_node("Dorf", level_count-1, parent_hash)
        else:
            child = None
        if child:
            node["children"].append(child)
    
    # Hash des Knotens = Hash des Namens + Hash der Kinder + Hash des Elternknotens
    node_data = {
        "name": node["name"],
        "children_hashes": [c["hash"] for c in node["children"]],
        "parent_hash": parent_hash
    }
    node["hash"] = hash_data(node_data)
    
    return node

# Erzeuge Baum mit Wurzel = Länder-Ebene (3 Ebenen tiefer)
tree = generate_node("Land", 3)
print(json.dumps(tree, indent=2))

File: blockchainPAdic/blockchainPAdic.hs
{-# LANGUAGE OverloadedStrings #-}
import Crypto.Hash.SHA256 (hash)
import qualified Data.ByteString.Char8 as BS
import Data.ByteString.Base16 (encode)
import System.Random

data Node = Node {
    name :: String,
    hashVal :: String,
    children :: [Node]
} deriving Show

sha256 :: String -> String
sha256 s = BS.unpack $ encode $ hash $ BS.pack s

generateName :: String -> IO String
generateName level = do
    n <- randomRIO (1,1000)
    return $ level ++ "_" ++ show n

generateNode :: String -> Int -> String -> IO Node
generateNode level depth parentHash = do
    n <- generateName level
    numChildren <- if depth > 0 then randomRIO (1,3) else return 0
    childNodes <- mapM (\_ -> 
        if depth == 3 then generateNode "Bundesland" (depth-1) parentHash
        else if depth == 2 then generateNode "Bezirk" (depth-1) parentHash
        else if depth == 1 then generateNode "Dorf" (depth-1) parentHash
        else return undefined
        ) [1..numChildren]
    let nodeHash = sha256 (n ++ parentHash ++ concatMap hashVal childNodes)
    return $ Node n nodeHash childNodes

printNode :: Node -> Int -> IO ()
printNode node indent = do
    putStrLn $ replicate (indent*2) ' ' ++ name node ++ " : " ++ hashVal node
    mapM_ (\c -> printNode c (indent+1)) (children node)

main :: IO ()
main = do
    tree <- generateNode "Land" 3 ""
    printNode tree 0
